### ISSUE TITLE
Enable syntax highlighting in JupyterLab by default

### ISSUE THREAD

**Issue by jamesdbrock** [gibiansky/IHaskell#1199]
I think Haskell syntax highlighting should be enabled by default. One way to do that would be to insert a switch in `kernel.json`, maybe like

```Haskell
++ ["--codemirror", "Haskell"]
```

after the `--ghclib` argument.

https://github.com/gibiansky/IHaskell/blob/ef698157f44960566687a308e3455b5ba031eb43/src/IHaskell/IPython.hs#L140-L144

But if we did this, then how would syntax highlighting be disabled?

Maybe there's a more natural way to accomplish this.

Reference: https://github.com/gibiansky/IHaskell/pull/1191

**Comment by jamesdbrock** [comment]
I have some observations to make about the current behavior of `ihaskell` Codemirror on the master branch:

1. `--codemirror haskell` will enable syntax highlighting in Lab.
2. `--codemirror Haskell` will enable syntax highlighting in Lab.
3. `--codemirror ihaskell` will not enable syntax highlighting in Lab.
4. In the classic Notebook no switch is needed, Codemirror is already enabled.

**Comment by jamesdbrock** [comment]
Soo... what's going on here is that JupyterLab now, by default, already sets notebook cells to

```
"language_info": {
   "codemirror_mode": "ihaskell",
```

But the `kernel.js` script for the special `"ihaskell"` multiplexing mode isn't working anymore.

**Comment by jamesdbrock** [comment]
I just realized that the special `"ihaskell"` Codemirror mode in `kernel.js` was never working in JupyterLab.

`kernel.js` is for classic notebook only. 

The commandline switch which @MMesch added in #1191 allows us to specify `"haskell"` mode for Codemirror, which works because Codemirror comes with `"haskell"` mode built-in. But that mode will highlight command code cells prefixed with `:` as if they were Haskell. Which is the special case that the special `"ihaskell"` mode addresses.

**Comment by jamesdbrock** [comment]
Apparently adding Codemirror modes to kernels in JupyterLab is a secret art, but here are some clues:

https://github.com/jupyterlab/jupyterlab/pull/9067
https://github.com/jupyterlab/jupyterlab/pull/9123
https://github.com/jupyterlab/jupyterlab/issues/10095
https://discourse.jupyter.org/t/how-to-find-the-codemirror-instance/9275

**Comment by jamesdbrock** [comment]
I think that the extension which we already have is the best way to add Codemirror modes to kernels in JupyterLab: https://github.com/gibiansky/IHaskell/blob/master/jupyterlab-ihaskell/src/codemirror-ihaskell.ts

**Comment by jamesdbrock** [comment]
What is this about?

> A kernel may specify a `codemirror-mode.js`

https://github.com/jupyterlab/jupyterlab/issues/3010

**Comment by jamesdbrock** [comment]
Also note the extension directory name was changed in https://github.com/gibiansky/IHaskell/commit/e108c9cd166777c5aeb15e6485a1bc49d6f82818

so IHaskell README needs update.

**Comment by vaibhavsagar** [comment]
I don't follow, the README specifies `jupyterlab-ihaskell`. Am I misunderstanding?

**Comment by jamesdbrock** [comment]
The top README: https://github.com/gibiansky/IHaskell/blob/master/README.md#stack-and-docker

**Comment by vaibhavsagar** [comment]
Updated.

**Comment by jamesdbrock** [comment]
Thx.

Do you have any thoughts on this issue, Vaibhav? I would like to make Haskell syntax highlighting in JupyterLab work automatically without making users install the labextension. But that doesn't currently work, and I don't see any way to improve on the current situation.

**Comment by krassowski** [comment]
Have you considered using [prebuilt extensions](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#prebuilt-extensions) to eliminate the need for Node.js and npm?

Your installer would just need to drop the pre-compiled `jupyterlab-ihaskell` in an [appropriated directory](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#distributing-prebuilt-extensions). [jupyterlab_robotmode](https://github.com/MarketSquare/jupyterlab_robotmode) uses this approach.

**Comment by vaibhavsagar** [comment]
I think this is the problem though, we still need an extra installation step for something that "just works" with the older Classic Notebook interface. I can't see a way around that given the JupyterLab architecture.

**Comment by vaibhavsagar** [comment]
Maybe we could get `ihaskell install` to copy a pre-compiled `jupyterlab-ihaskell` to the right place?

**Comment by vaibhavsagar** [comment]
I tried using a prebuilt version and it does seem easy to install but I ran into this issue: https://discourse.jupyter.org/t/whats-wrong-with-the-way-im-using-codemirror-in-this-prebuilt-version-of-jupyterlab-ihaskell/10539, does anyone have any ideas?

**Comment by vaibhavsagar** [comment]
Okay, I've fixed the issue and the prebuilt extension is available [here](https://github.com/gibiansky/IHaskell/tree/master/jupyterlab-ihaskell/labextension), you should be able to copy it to the `labextensions/jupyterlab-ihaskell` directory in your `jupyter --data-dir` (e.g. for me that's `~/.local/share/jupyter/labextensions/jupyterlab-ihaskell`). Unfortunately the files contain more than one dot in their filename so because of https://github.com/haskell/cabal/issues/5057 I can't include them in IHaskell without dropping support for Cabal < 3.0. I'll keep thinking about this but for now I can't see a straightforward way past this issue.

**Comment by vaibhavsagar** [comment]
Whoops, didn't mean to close this.

**Comment by vaibhavsagar** [comment]
I think a potential solution is to use a custom webpack config to remove one of the dots in the filename so we can then use the currently support file globbing functionality. I've asked about this here but any other suggestions are welcome: https://discourse.jupyter.org/t/custom-webpack-configuration-for-jupyterlab-prebuilt-extension-with-special-filename-formatting/10570.

**Comment by vaibhavsagar** [comment]
I've implemented this and intend to merge soon.

**Comment by vaibhavsagar** [comment]
Fixed by #1270.
